README.md
# TYPO3 Extension mask_export

[![Latest Stable Version](https://img.shields.io/packagist/v/ichhabrecht/mask-export.svg)](https://packagist.org/packages/ichhabrecht/mask-export)
[![Build Status](https://img.shields.io/travis/IchHabRecht/mask_export/main.svg)](https://travis-ci.org/IchHabRecht/mask_export)
[![StyleCI](https://styleci.io/repos/63010277/shield?branch=main)](https://styleci.io/repos/63010277)

Want to create new content elements the easy way?

Use [mask](http://mask.webprofil.at) and its drag & drop wizard to create content elements the easy way.

Use mask_export to export the content elements into an own extension.

## Why

A content element needs some TCA information, TypoScript and database configuration and template files.
Actually there is no need to depend on any third party extension when dealing with content elements.

But you don't want to deal with different information in different files and folders.
You want to be able to concentrate on what is necessary for the user, not how it has to be implemented in your TYPO3 CMS.

By using mask and mask_export you can do exactly this! Simply create your own content elements by using a drag & drop wizard.
Add fields, repeating items, nested content elements within minutes.

## Why not

Mask offers an easy way to create content elements but has some disadvantage when you need to take care about performance.
Especially the frontend rendering can take at lot of time for uncached pages.
Instead of relying on TYPO3 CMS core rendering, all elements are rendered by an own Extbase Controller.
As this was needed to support former version of TYPO3 CMS, it was deprecated with the introduction of fluid_styled_content and the concepts of DataProcessors in TYPO 7 LTS.

This is what mask_export is developed for. It takes the content element information from the mask configuration and generates the needed
code to get those elements to work with pure TYPO3 CMS core functionality out of the box.
It bundles all necessary information into an own extension that can be installed and used in every other TYPO3 CMS system.

## Installation

Simply install mask and mask_export with Composer or the Extension Manager.

`composer require ichhabrecht/mask-export`

## Usage

- use the mask wizard to configure own content elements
- change to tab "Code Export"
- if you like change the extension key, the default one is *my_mask_export*
- either install or download your extension

## Best practise

It is recommended to **not touch** the generated export extension.
Instead you should add necessary changes and your own templates into a [site package](https://sitepackagebuilder.com/).

This ensures you can edit your content elements within the mask wizard (add further content elements, change settings)
and still be able to use the new extension version in your existing system.

You can find some common configuration in the [my_maskexport_sitepackage](https://github.com/IchHabRecht/my_maskexport_sitepackage)
example site package.

Furthermore you can refer to the slides [CCE (Custom Content Elements) - Best Practice ](https://de.slideshare.net/cpsitgmbh/cce-custom-content-elements-best-practice)
for additional information. 

## Community

- Thanks to [Marcus Schwemer](https://twitter.com/MarcusSchwemer) who wrote about mask_export in his blog [TYPO3worx](https://typo3worx.eu/2018/03/eight-typo3-extensions-making-developers-happy/)
- Thanks to [Thomas Löffler](https://spooner-web.de) for his ongoing support as [Patron](https://www.patreon.com/IchHabRecht)


ChangeLog
2022-05-05         [RELEASE] Release of mask_export 4.0.2 (Nicole Cordes)
2022-04-29 db7a3c1 [BUGFIX] Ensure compatibility of the HTML attributes for all supported TYPO3 versions (Noel Kayabasli)
2022-04-29 f624a89 [TASK] Change the label and description of the backend module (Noel Kayabasli)
2022-04-29 8ab468d [BUGFIX] Fix HTML attributes in the code preview panels (Noel Kayabasli)
2022-04-29 f44cbd8 [TASK] Change the label and description of the module (Noel Kayabasli)

2022-04-28         [RELEASE] Release of mask_export 4.0.1 (Nicole Cordes)
2022-04-28 9c6e10b [BUGFIX] Prevent PHP (>=8.0) warning (Nicole Cordes)
2022-04-28 da56348 [TASK] Update .gitignore file (Nicole Cordes)

2022-04-25         [RELEASE] Release of mask_export 4.0.0 (Nicole Cordes)
2022-04-24 c507e00 [TASK] Migrate PHPUnit assert functions (Nicole Cordes)
2022-04-24 b299a74 [TASK] Add compatibility to PHP 8.1 (Nicole Cordes)
2022-04-19 0cd56b4 [BUGFIX] Fix failing functional tests (Nicole Cordes)
2022-04-19 1c7cf3e [TASK] Add missing static closures in export code (Nicole Cordes)
2022-04-19 71fe4a5 [TASK] Add missing static closures (Nicole Cordes)
2022-04-02 c4fcceb [TASK] Define inline functions static (Simon Schaufelberger)
2022-04-17 a46b7b7 [BUGFIX] Fix sonar tests for Travis CI (Nicole Cordes)
2022-04-16 1357a0d [FEATURE] Add compatibility to TYPO3 11.5 (Nicole Cordes)
2022-04-16 91a9ede [TASK] Enforce phpunit/phpunit ^8.5 (Nicole Cordes)
2022-04-16 cbd9048 [TASK] Use svg icon as default element icon (Nicole Cordes)
2022-04-13 dc2ea35 [TASK] Use correct regular expression for item labels (Nicole Cordes)
2022-04-13 8919179 [BUGFIX] Use correct extension key in functional tests (Nicole Cordes)
2022-04-13 e825127 [BUGFIX] Prevent core fields in export (Nicole Cordes)
2022-04-13 3972a45 [BUGFIX] Store configuration as ElementDefinition (Nicole Cordes)
2022-04-13 a1ff531 [TASK] Add support for mask 7.1 (Nicole Cordes)
2022-04-12 7c1ec56 [TASK] Start 4.x development (Nicole Cordes)
2022-04-11 eb66ef8 [TASK] Switch branch name to main (Nicole Cordes)
2022-04-11 2012109 [BUGFIX] Fix failing tests for mask/mask: ^6.0 (Nicole Cordes)
2021-11-11 9c1260d [BUGFIX] Use correct file name for language file (Nicole Cordes)
2021-11-11 6416125 [BUGFIX] Update path to language file (Georg Ringer)
2021-03-13 8a63b2a [BUGFIX] Fix default TCA iconfile (Rémy DANIEL)
2021-03-16 3497c61 [BUGFIX] Remove "use" statement from ext_tables.php (Nicole Cordes)

2020-10-27         [RELEASE] Release of mask_export 3.0.4 (Nicole Cordes)
2020-10-27 5811870 [TASK] Update supporting files to EXT:my_skeleton 0.5.1 (Nicole Cordes)
2020-10-27 a8ce63e [BUGFIX] Prepare bugfix release (Nicole Cordes)
2020-10-27 e0b7a55 [TASK] Make HtmlCodeGenerator compatible with older mask versions (Nikita Hovratov)
2020-10-27 a947ec7 [TASK] Update HtmlCodeGenerator for mask version 6 (Nikita Hovratov)

2020-10-22         [RELEASE] Release of mask_export 3.0.3 (Nicole Cordes)
2020-10-22 70bee19 [TASK] Add support for mask 6.0 (Nicole Cordes)
2020-10-22 c7d7d3b [TASK] Update supporting files to EXT:my_skeleton 0.4.0 (Nicole Cordes)

2020-07-23         [RELEASE] Release of mask_export 3.0.2 (Nicole Cordes)
2020-07-20 1e2cd58 [BUGFIX] Remove dependency to AbstractCodeGenerator (Nicole Cordes)

2020-06-07         [RELEASE] Release of mask_export 3.0.1 (Nicole Cordes)
2020-06-05 d1d4032 [BUGFIX] Remove configuration options from export handling (Nicole Cordes)
2020-05-30 8d89754 [BUGFIX] Remove dev-master tests from Travis CI configuration (Nicole Cordes)
2020-05-27 71b1418 [BUGFIX] Fix functional tests on Travis CI (Nicole Cordes)
2020-05-22 582917a [BUGFIX] Enforce phpunit/phpunit ^7.5 (Nicole Cordes)

2020-05-18         [RELEASE] Release of mask_export 3.0.0 (Nicole Cordes)
2020-05-16 9ef69ee [FEATURE] Remove original mask elements from TCA (Nicole Cordes)
2020-05-14 ecb11df [FEATURE] Add ability to select content elements (Nicole Cordes)
2020-05-12 b9b4d2e [TASK] Remove support for deprecated configuration in backend preview (Nicole Cordes)
2020-05-12 54143cf [TASK] Beautify php code output (Nicole Cordes)
2020-05-12 b3f8f49 [TASK] Simplify ext_tables.sql output (Nicole Cordes)
2020-05-11 499b60b [TASK] Use correct file extensions for configuration files (Nicole Cordes)
2020-05-11 6d6762a [TASK] Remove unsupported replacement in composer.json (Nicole Cordes)
2020-05-11 7460476 [TASK] Use svg extension icon (Nicole Cordes)
2020-05-11 b51fa64 [BUGFIX] Fix functional tests (Nicole Cordes)
2020-05-10 f7f98e2 [FEATURE] Show export as own backend module (Nicole Cordes)
2020-05-10 25a93ce [FEATURE] Add compatibility to TYPO3 10.4 (Nicole Cordes)
2020-05-10 949a06c [TASK] Update supporting files to EXT:my_skeleton 0.3.2 (Nicole Cordes)
2020-05-10 03c2d56 [TASK] Start 3.x development (Nicole Cordes)

2019-08-20         [RELEASE] Release of mask_export 2.4.0 (Nicole Cordes)
2019-08-20 27d942d [TASK] Update README.md (Nicole Cordes)
2019-08-03 47fa5fd [TASK] Update supporting files to EXT:my_skeleton 0.2.1 (Nicole Cordes)
2019-07-03 b8df9f8 [BUGFIX] Add indexes to tt_content for parent fields (Nicole Cordes)
2019-05-22 cfcaf4a [BUGFIX] Compare JSON strings instead of file content (Nicole Cordes)
2019-05-22 ffc920c [BUGFIX] Ensure unchanged mask configuration for export (Nicole Cordes)
2019-05-22 bac264e [FEATURE] Move extension and vendor name to JSON test files (Nicole Cordes)
2019-05-22 c6237d4 [FEATURE] Allow name configuration in mask.json (Nicole Cordes)

2019-05-20         [RELEASE] Release of mask_export 2.3.0 (Nicole Cordes)
2019-03-25 fef554d [FEATURE] Add unit and functional tests for PhpFile flags (Nicole Cordes)
2019-03-25 8085238 [FEATURE] Introduce FlagResolver api (Nicole Cordes)
2019-03-24 58d5a29 [FEATURE] Improve PHP file output (Nicole Cordes)
2019-03-23 eb7b267 [BUGFIX] Support non-array colPos configuration (Nicole Cordes)
2019-03-22 342f97b [BUGFIX] Ensure only labels for mask fields are exported (Nicole Cordes)


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "mask_export".
 *
 * Auto generated 05-05-2022 18:31
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array (
  'title' => 'Mask Export',
  'description' => 'Export your mask elements as extension',
  'category' => 'module',
  'author' => 'Nicole Cordes',
  'author_email' => 'typo3@cordes.co',
  'author_company' => 'CPS-IT GmbH | biz-design.biz',
  'state' => 'stable',
  'version' => '4.0.2',
  'constraints' => 
  array (
    'depends' => 
    array (
      'typo3' => '9.5.0-11.5.99',
      'mask' => '4.0.0-7.99.99',
    ),
    'conflicts' => 
    array (
    ),
    'suggests' => 
    array (
    ),
  ),
  '_md5_values_when_last_written' => 'a:72:{s:9:"ChangeLog";s:4:"8ca9";s:7:"LICENSE";s:4:"b234";s:9:"README.md";s:4:"8458";s:13:"composer.json";s:4:"7cf1";s:13:"composer.lock";s:4:"5bf8";s:12:"ext_icon.png";s:4:"03b4";s:14:"ext_tables.php";s:4:"7ec9";s:16:"phpunit.xml.dist";s:4:"041c";s:24:"sonar-project.properties";s:4:"7de6";s:39:"Classes/Aggregate/AbstractAggregate.php";s:4:"7641";s:52:"Classes/Aggregate/AbstractInlineContentAggregate.php";s:4:"b29c";s:48:"Classes/Aggregate/AbstractOverridesAggregate.php";s:4:"1273";s:41:"Classes/Aggregate/AggregateCollection.php";s:4:"4341";s:45:"Classes/Aggregate/BackendPreviewAggregate.php";s:4:"df9f";s:49:"Classes/Aggregate/ContentElementIconAggregate.php";s:4:"8f1b";s:47:"Classes/Aggregate/ContentRenderingAggregate.php";s:4:"dc5a";s:53:"Classes/Aggregate/ExtensionConfigurationAggregate.php";s:4:"2fd7";s:49:"Classes/Aggregate/InlineContentCTypeAggregate.php";s:4:"8691";s:50:"Classes/Aggregate/InlineContentColPosAggregate.php";s:4:"a49f";s:44:"Classes/Aggregate/LanguageAwareInterface.php";s:4:"4cc3";s:40:"Classes/Aggregate/LanguageAwareTrait.php";s:4:"82e9";s:54:"Classes/Aggregate/NewContentElementWizardAggregate.php";s:4:"8372";s:39:"Classes/Aggregate/PhpAwareInterface.php";s:4:"bfa9";s:35:"Classes/Aggregate/PhpAwareTrait.php";s:4:"277d";s:49:"Classes/Aggregate/PlainTextFileAwareInterface.php";s:4:"0767";s:45:"Classes/Aggregate/PlainTextFileAwareTrait.php";s:4:"7c14";s:39:"Classes/Aggregate/SqlAwareInterface.php";s:4:"e57c";s:35:"Classes/Aggregate/SqlAwareTrait.php";s:4:"5747";s:34:"Classes/Aggregate/TcaAggregate.php";s:4:"1231";s:35:"Classes/Aggregate/TcaAwareTrait.php";s:4:"9467";s:49:"Classes/Aggregate/TtContentOverridesAggregate.php";s:4:"bec2";s:51:"Classes/CodeGenerator/BackendFluidCodeGenerator.php";s:4:"c305";s:43:"Classes/CodeGenerator/HtmlCodeGenerator.php";s:4:"0326";s:39:"Classes/Controller/ExportController.php";s:4:"7831";s:49:"Classes/FileCollection/AbstractFileCollection.php";s:4:"0184";s:41:"Classes/FileCollection/FileCollection.php";s:4:"d9ab";s:49:"Classes/FileCollection/LanguageFileCollection.php";s:4:"76eb";s:44:"Classes/FileCollection/PhpFileCollection.php";s:4:"2f5c";s:50:"Classes/FileCollection/PlainTextFileCollection.php";s:4:"f059";s:44:"Classes/FileCollection/SqlFileCollection.php";s:4:"52fb";s:45:"Classes/FlagResolver/AbstractFlagResolver.php";s:4:"82d8";s:38:"Classes/FlagResolver/FlagInterface.php";s:4:"8c65";s:46:"Classes/FlagResolver/FlagResolverInterface.php";s:4:"226d";s:44:"Classes/FlagResolver/PhpFileFlagResolver.php";s:4:"9ace";s:56:"Classes/FlagResolver/PhpFileFlag/ClosureFunctionFlag.php";s:4:"43ae";s:57:"Classes/FlagResolver/PhpFileFlag/DefinedTypo3ModeFlag.php";s:4:"c44a";s:52:"Classes/FlagResolver/PhpFileFlag/PhpStartTagFlag.php";s:4:"58a3";s:44:"Resources/Private/Language/locallang_mod.xlf";s:4:"d13f";s:38:"Resources/Private/Layouts/Default.html";s:4:"8382";s:44:"Resources/Private/Templates/Export/List.html";s:4:"6995";s:36:"Resources/Public/Icons/Extension.svg";s:4:"d494";s:38:"Resources/Public/JavaScript/Toggler.js";s:4:"1880";s:64:"Tests/Functional/Controller/AbstractExportControllerTestCase.php";s:4:"cbb1";s:67:"Tests/Functional/Controller/BackendPreview/ExportControllerTest.php";s:4:"3b1a";s:66:"Tests/Functional/Controller/Configuration/ExportControllerTest.php";s:4:"80f0";s:65:"Tests/Functional/Controller/DataProvider/ExportControllerTest.php";s:4:"46e5";s:68:"Tests/Functional/Controller/FlagResolver/PhpFileFlagResolverTest.php";s:4:"2394";s:66:"Tests/Functional/Controller/MaskConfiguration/EmptyColumnsTest.php";s:4:"01b0";s:76:"Tests/Functional/Controller/MaskConfiguration/EmptyMaskConfigurationTest.php";s:4:"49d0";s:76:"Tests/Functional/Controller/NewContentElementWizard/ExportControllerTest.php";s:4:"fc87";s:63:"Tests/Functional/Controller/Ressources/ExportControllerTest.php";s:4:"f54f";s:63:"Tests/Functional/Controller/TypoScript/ExportControllerTest.php";s:4:"8152";s:57:"Tests/Functional/Fixtures/Configuration/mask-default.json";s:4:"68be";s:63:"Tests/Functional/Fixtures/Configuration/mask-empty-columns.json";s:4:"9c6c";s:44:"Tests/Functional/Fixtures/Database/pages.xml";s:4:"5d05";s:47:"Tests/Functional/Fixtures/Database/sys_file.xml";s:4:"fac7";s:49:"Tests/Functional/Fixtures/Database/tt_content.xml";s:4:"9b39";s:77:"Tests/Functional/Fixtures/Database/tx_maskexampleexport_additionalcontent.xml";s:4:"df9e";s:52:"Tests/Functional/Fixtures/Templates/Export/List.html";s:4:"d41d";s:74:"Tests/Functional/Fixtures/Templates/Preview/ce_nested-content-elements.png";s:4:"5ef4";s:62:"Tests/Functional/Fixtures/Templates/Preview/simple-element.png";s:4:"78c9";s:51:"Tests/Unit/FileCollection/PhpFileCollectionTest.php";s:4:"c566";}',
);

